reconcile/ocm_machine_pools.py

```diff
@@ -1,137 +1,332 @@
+                    diffs.append(
+                        PoolHandler(
+                            action="create",
+                            pool=NodePool.create_from_gql(
+                                pool=desired_machine_pool,
+                                cluster=cluster_name,
+                            ),
+                        )
+                    )
+                else:
+                    diffs.append(
+                        PoolHandler(
+                            action="create",
+                            pool=MachinePool.create_from_gql(
+                                pool=desired_machine_pool,
+                                cluster=cluster_name,
+                            ),
+                        )
+                    )
+                    continue
+            if current_machine_pool[0].has_diff(desired_machine_pool):
+                invalid_diff = current_machine_pool[0].invalid_diff(
+                    desired_machine_pool
+                )
+                if invalid_diff:
+                    errors.append(
+                        InvalidUpdateError(
+                            f"can not update {invalid_diff} for existing machine pool"
+                        )
+                    )
+                else:
+                    diffs.append(
+                        PoolHandler(
+                            action="update",
+                            pool=current_machine_pool[0],
+                        )
+                    )
 
-    return diffs, err
+    for cluster_name, current_data in current_state.items():
+        for current_machine_pool in current_data["machine_pools"]:
+            if f"{current_machine_pool.id}-{cluster_name}" not in all_desired_pools:
+                diffs.append(
+                    PoolHandler(
+                        action="delete",
+                        pool=current_machine_pool,
+                    )
+                )
+    return diffs, errors
 
 
-def act(dry_run, diffs, ocm_map):
+def act(dry_run: bool, diffs: Iterable[PoolHandler], ocm_map: OCMMap) -> None:
     for diff in diffs:
-        action = diff.pop("action")
-        cluster = diff.pop("cluster")
-        logging.info([action, cluster, diff["id"]])
+        logging.info([diff.action, diff.pool.cluster.name, diff.pool.id])
         if not dry_run:
-            ocm = ocm_map.get(cluster)
-            if action == "create":
-                ocm.create_machine_pool(cluster, diff)
-            elif action == "update":
-                ocm.update_machine_pool(cluster, diff)
-            elif action == "delete":
-                ocm.delete_machine_pool(cluster, diff)
+            ocm = ocm_map.get(diff.pool.cluster.name)
+            diff.act(dry_run, ocm)
+
 
+def _cluster_is_compatible(cluster: ClusterV1) -> bool:
+    return (
+        cluster.ocm is not None
+        and cluster.machine_pools is not None
+        # and not cluster.spec.hypershift
+    )
```

Comment:
don't you need to check if `spec` is set?

Reply:
Good point, but that line was for testing only.